File: DSA_CSharp/MergeSortExample.cs
using System;

namespace DSA
{
    public class MergeSortExample
    {
        // Merge Sort = Divide and Conquer algorithm that splits an array into smaller subarrays,
        //              sorts those subarrays, and then merges them back together
        //              More efficient than simpler algorithms like Bubble Sort, Selection Sort, and Insertion Sort
        //              Runtime complexity: O(n log n) in all cases (best, average, worst)
        //              Advantages:
        //              Consistent O(n log n) performance
        //              Stable sort (preserves the relative order of equal elements)
        //              Well-suited for large data sets and linked lists
        //              Disadvantages:
        //              Requires additional memory for temporary arrays during the merge process
        //              Generally slower for small data sets compared to simpler algorithms
        public MergeSortExample()
        {


            int[] array = { 9, 5, 7, 3, 2, 1, 6, 4, 8 };
            Console.WriteLine("Unsorted Array: " + string.Join(", ", array));

            MergeSort(array);

            Console.WriteLine("Sorted Array: " + string.Join(", ", array));
        }

        private static void MergeSort(int[] array)
        {
            int length = array.Length;
            if (length <= 1) return; //base case 

            int middle = length / 2;
            int[] leftArray = new int[middle];
            int[] rightArray = new int[length - middle];

            int i = 0; //left array
            int j = 0; //right array

            for (; i < length; i++)
            {
                if (i < middle)
                {
                    leftArray[i] = array[i];
                }
                else
                {
                    rightArray[j] = array[i];
                    j++;
                }
            }

            MergeSort(leftArray);
            MergeSort(rightArray);
            Merge(leftArray, rightArray, array);
        }

        private static void Merge(int[] leftArray, int[] rightArray, int[] array)
        {
            int leftSize = array.Length / 2;
            int rightSize = array.Length - leftSize;
            int i = 0, l = 0, r = 0; //indices

            //check the conditions for merging
            while (l < leftSize && r < rightSize)
            {
                if (leftArray[l] < rightArray[r])
                {
                    array[i] = leftArray[l];
                    i++;
                    l++;

                }
                else
                {
                    array[i] = rightArray[r];
                    i++;
                    r++;
                }
            }

            while (l < leftSize)
            {
                array[i] = leftArray[l];
                i++;
                l++;
            }
            while (r < rightSize)
            {
                array[i] = rightArray[r];
                i++;
                r++;
            }

        }
    }
}

File: DSA_CPP/MergeSortExample.cpp
#include "MergeSortExample.h"
#include <vector>
#include <iostream>

void MergeSortExample::demonstrate()
{
    std::cout << "\n=== Merge Sort Example ===\n";
    std::cout << "Merge Sort: Divide-and-conquer stable sorting algorithm\n";
    std::cout << "Time Complexity: O(n log n) - Best, Average, and Worst case\n";
    std::cout << "Space Complexity: O(n) - Additional space for merging\n";
    std::cout << "Stable: Yes - Maintains relative order of equal elements\n\n";

    // Test with different arrays
    std::vector<int> arr1 = {64, 34, 25, 12, 22, 11, 90};
    std::cout << "Example 1 - Random array:\n";
    demonstrateMergeSort(arr1);

    std::vector<int> arr2 = {5, 2, 4, 6, 1, 3};
    std::cout << "\nExample 2 - Small array:\n";
    demonstrateMergeSort(arr2);

    std::vector<int> arr3 = {1, 2, 3, 4, 5};
    std::cout << "\nExample 3 - Already sorted:\n";
    demonstrateMergeSort(arr3);

    std::vector<int> arr4 = {5, 4, 3, 2, 1};
    std::cout << "\nExample 4 - Reverse sorted:\n";
    demonstrateMergeSort(arr4);

    std::cout << "\n--- Merge Sort Properties ---\n";
    std::cout << "✓ Consistent O(n log n) performance\n";
    std::cout << "✓ Stable sorting algorithm\n";
    std::cout << "✓ Predictable performance - no worst case degradation\n";
    std::cout << "✓ Good for large datasets\n";
    std::cout << "✗ Requires additional O(n) space\n";
    std::cout << "✗ Not in-place sorting\n";

    std::cout << "\n--- Use Cases ---\n";
    std::cout << "• Large datasets requiring guaranteed O(n log n) performance\n";
    std::cout << "• When stability is important (maintaining relative order)\n";
    std::cout << "• External sorting (when data doesn't fit in memory)\n";
    std::cout << "• When consistent performance is more important than space\n";
}

void MergeSortExample::demonstrateMergeSort(std::vector<int> arr)
{
    std::cout << "Original array: ";
    printArray(arr);

    mergeSort(arr, 0, arr.size() - 1);

    std::cout << "Sorted array:   ";
    printArray(arr);
}

void MergeSortExample::mergeSort(std::vector<int> &arr, int left, int right)
{
    if (left < right)
    {
        // Find the middle point
        int mid = left + (right - left) / 2;

        // Sort first and second halves
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        // Merge the sorted halves
        merge(arr, left, mid, right);
    }
}

void MergeSortExample::merge(std::vector<int> &arr, int left, int mid, int right)
{
    // Calculate sizes of the two subarrays
    int n1 = mid - left + 1;
    int n2 = right - mid;

    // Create temporary arrays
    std::vector<int> leftArr(n1);
    std::vector<int> rightArr(n2);

    // Copy data to temporary arrays
    for (int i = 0; i < n1; i++)
        leftArr[i] = arr[left + i];
    for (int j = 0; j < n2; j++)
        rightArr[j] = arr[mid + 1 + j];

    // Merge the temporary arrays back into arr[left..right]
    int i = 0;    // Initial index of first subarray
    int j = 0;    // Initial index of second subarray
    int k = left; // Initial index of merged subarray

    while (i < n1 && j < n2)
    {
        if (leftArr[i] <= rightArr[j])
        {
            arr[k] = leftArr[i];
            i++;
        }
        else
        {
            arr[k] = rightArr[j];
            j++;
        }
        k++;
    }

    // Copy the remaining elements of leftArr[], if any
    while (i < n1)
    {
        arr[k] = leftArr[i];
        i++;
        k++;
    }

    // Copy the remaining elements of rightArr[], if any
    while (j < n2)
    {
        arr[k] = rightArr[j];
        j++;
        k++;
    }
}

void MergeSortExample::printArray(const std::vector<int> &arr)
{
    for (size_t i = 0; i < arr.size(); i++)
    {
        std::cout << arr[i];
        if (i < arr.size() - 1)
            std::cout << ", ";
    }
    std::cout << std::endl;
}